### PR TITLE
docs: historic spec updates

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

Updates the historacle spec to allow for use by both the leverage module & ibc rate limiting

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by adamewozniak** [comment]
@robert-zaremba / @toteki can I please get reviews here? iirc this is what we agreed upon on our last meeting, just want to be sure 🙏

File: docs/design_docs/011-historacle-pricing.md [hunk 1]
```diff
@@ -34,21 +35,31 @@ Historacle pricing will provide an API for the leverage module to tell when pric
...
 
-The `x/leverage` module will have access to the following `keeper` functions from the `x/oracle` module:
+Modules will have access to the following `keeper` functions from the `x/oracle` module:
+
 - `HistoricMedian(denom) (sdk.Dec, error)` returns the median price of an asset in the last `Pruning Period`
 - `WithinHistoricDeviation(denom) (bool, error)` returns whether or not the current price of an asset is within the `Standard Deviation around the Median`.
-- `IsHistoricAsset(denom string) bool`, returns `true` if `denom` is a historacle asset, and returns `false` if it is not.
+- `MedianOfMedians(denom string, blockNum int) sdk.Dec` returns the Median of the all the Medians recorded within the past `blockNum`.
+- `AverageOfMedians(denom string, blockNum int) sdk.Dec` returns the Average of all the Medians recorded within the past `blockNum`.
+- `MaxMedian(denom string, blockNum int) sdk.Dec` returns the Maximum of all the Medians recorded within the past `blockNum`.
```
Comment (toteki) [inline docs/design_docs/011-historacle-pricing.md:62]:
```suggestion
- `MaxMedian(denom string, blockNum int) sdk.Dec` returns the Maximum of all the Medians recorded within the past `blockNum`.
- `MinMedian(denom string, blockNum int) sdk.Dec` returns the Minimum of all the Medians recorded within the past `blockNum`.
```
For completeness - this might be used in a "min collateral price, max borrow price" implementation or similar.

File: docs/design_docs/011-historacle-pricing.md [hunk 2]
```diff
@@ -96,7 +107,7 @@ This will not introduce a new module, and it is relatively backwards compatible.
 
 ## Comments
 
-Originally a 24-hour median was thought to be effective enough to defend against this type of attack, but our modeling shows that we would need 30 days for adequate protection.
+Currently we are planning on keeping the last 6 hours or so of medians, and a longer historic price period. This way each median can be calculated based off of the last 1 hour or 12 hours, and we can perform computations over the last x amount of medians.
```
Comment (toteki) [inline docs/design_docs/011-historacle-pricing.md:110]: Might clarify what's happening with the 1-12 hours

File: docs/design_docs/011-historacle-pricing.md [hunk 3]
```diff
@@ -34,21 +35,31 @@ Historacle pricing will provide an API for the leverage module to tell when pric
...
+
+### Epochs
+
 We define three epoch periods, during which additional computation will be performed:
-- `Stamp Period`: duration during which the `x/oracle` module will now "stamp" the set of exchange rates in the state machine until a `Pruning Period` has passed (30 days).
-- `Pruning Period`: duration after which the `x/oracle` module will begin to prune expired historic prices.
-- `Median Period`: will determine how often the Median and the `Standard Deviation around the Median` are calculated, which will also be stored in the state machine.
 
-These values are stored in state in order to avoid the `x/leverage` module from having to calculate them while making decisions around allowable positions for users to take.
+- `Historic Stamp Period`: duration during which the `x/oracle` module will now "stamp" the set of exchange rates in the state machine until a `Pruning Period` has passed (30 days).
```
Comment (toteki) [inline docs/design_docs/011-historacle-pricing.md:44]: This sounds like we're keeping the once-per-5-block stamps for 30 days. I thought we were clearing every 6hrs

File: docs/design_docs/011-historacle-pricing.md (hunk 1 above)
Comment (adamewozniak) [reply docs/design_docs/011-historacle-pricing.md:62]: Added 👍

File: docs/design_docs/011-historacle-pricing.md (hunk 2 above)
Comment (adamewozniak) [reply docs/design_docs/011-historacle-pricing.md]: Clarified that the intention is for the "max medians" and "max price stamps" to be agnostic of one another

File: docs/design_docs/011-historacle-pricing.md (hunk 3 above)
Comment (adamewozniak) [reply docs/design_docs/011-historacle-pricing.md]: Clarified this bit.

Tldr; we'll stamp as often as we like (maybe every 100 blocks?) until a `Max Historic Prices` is met. So, you can stamp every 100 blocks, and every stamp will be valid for 6 hours.

File: docs/design_docs/011-historacle-pricing.md [hunk 4]
```diff
@@ -34,21 +35,39 @@ Historacle pricing will provide an API for the leverage module to tell when pric
...
-The `x/leverage` module will have access to the following `keeper` functions from the `x/oracle` module:
+Modules will have access to the following `keeper` functions from the `x/oracle` module:
+
 - `HistoricMedian(denom) (sdk.Dec, error)` returns the median price of an asset in the last `Pruning Period`
 - `WithinHistoricDeviation(denom) (bool, error)` returns whether or not the current price of an asset is within the `Standard Deviation around the Median`.
-- `IsHistoricAsset(denom string) bool`, returns `true` if `denom` is a historacle asset, and returns `false` if it is not.
+- `MedianOfMedians(denom string, blockNum int) sdk.Dec` returns the Median of the all the Medians recorded within the past `blockNum`.
+- `AverageOfMedians(denom string, blockNum int) sdk.Dec` returns the Average of all the Medians recorded within the past `blockNum`.
+- `MaxMedian(denom string, blockNum int) sdk.Dec` returns the Maximum of all the Medians recorded within the past `blockNum`.
+- `MaxMedian(denom string, blockNum int) sdk.Dec` returns the Minimum of all the Medians recorded within the past `blockNum`.
```
Comment (toteki) [inline docs/design_docs/011-historacle-pricing.md:70]:
```suggestion
- `MinMedian(denom string, blockNum int) sdk.Dec` returns the Minimum of all the Medians recorded within the past `blockNum`.
```

Comment (adamewozniak) [reply docs/design_docs/011-historacle-pricing.md]: thank u < 3

**Review by toteki — APPROVED** [review]
This seems like it covers everything.

**Comment by adamewozniak** [comment]
ok, I am merging this & moving forward
cc @rbajollari
